src/env.rs

```diff
@@ -604,30 +604,26 @@ pub unsafe fn get_cur_frame_segid(
     mut cur_seg_map: *const u8,
     stride: ptrdiff_t,
 ) -> (u8, u8) {
-    cur_seg_map = cur_seg_map.offset(bx as isize + by as isize * stride);
-    if have_left && have_top {
-        let l = *cur_seg_map.offset(-1);
-        let a = *cur_seg_map.offset(-stride as isize);
-        let al = *cur_seg_map.offset(-(stride + 1) as isize);
-        let seg_ctx = if l == a && al == l {
-            2
-        } else if l == a || al == l || a == al {
-            1
-        } else {
-            0
-        };
-        let seg_id = if a == al { a } else { l };
-        (seg_id, seg_ctx)
-    } else {
-        let seg_ctx = 0;
-        let seg_id = if have_left {
-            *cur_seg_map.offset(-1)
-        } else if have_top {
-            *cur_seg_map.offset(-stride as isize)
-        } else {
-            0
-        };
-        (seg_id, seg_ctx)
+    let negative_adjustment = have_left as usize + have_top as usize * stride.unsigned_abs();
+    cur_seg_map = cur_seg_map
+        .add((bx as isize + by as isize * stride - negative_adjustment as isize) as usize);
```

Comment:
@fbossen, when `stride` is negative, is `by` also going to be negative?  Because if `by` is positive and `stride` is negative, this `add` is probably going to be negative (ignoring the `as usize` at the end).  But `fn get_cur_frame_segid` is always called with `f.cur_segmap`, which is always the start of an allocation, so it can't be negatively offset.  So is `stride` always positive here, or is `by` negative when `stride` is negative?

Reply:
Looks like `stride` can't be negative here. The function is called with `stride =  f.b4_stride = (f.bw + 31 & !31) as ptrdiff_t;` (`f.bw` is positive).

Reply:
Awesome!  Good to know `f.bw` is positive.  That helps a lot.  I'll simply things then.